infra/Pos/Infra/Semaphore.hs

```diff
@@ -0,0 +1,80 @@
+-- | 'BlkSemaphore' primitive used for synchronization of threads in
+-- the application.
+
+module Pos.Infra.Semaphore
+       ( BlkSemaphore (..)
+       , withBlkSemaphore
+       , withBlkSemaphoreIgnoreTip
+       , withBlkSemaphore_
+       ) where
+
+import           Universum
+
+import           Control.Monad.Catch (MonadMask, bracketOnError)
+
+import           Pos.Core            (HeaderHash)
+import           Pos.Util.Util       (HasLens, lensOf)
+
+-- | A simple wrapper over 'MVar' which stores 'HeaderHash' (our
+-- current tip) and is taken whenever we want to update GState or
+-- other data dependent on GState.
+newtype BlkSemaphore = BlkSemaphore
+    { unBlkSemaphore :: MVar HeaderHash
+    }
+
+-- TODO: use 'withMVar' instead ('MonadThrow'-based).
```

Comment:
I'm not sure I understand this TODO, can you explain?

`withMVar` does essentially the same using a `bracketOnError` pattern.

Reply:
Ideally I would like to use a library function instead of defining it manually. There is `withMVar` in `base` which works in `IO` only, so it's not suitable. There is `withMVar` in `lifted-base`, but it's based on `MonadBaseControl IO`, while majority of our code uses `MonadMask` for these purposes. So I would like to define `withMVar` with `MonadMask` constraint somewhere in `Pos.Util` and use it here. I also need this version of `withMVar` for another issue I am working on.
I can also use `MonadBaseControl IO` version, but anyway, I want this logic with bracket to be defined in a more common place.
Also `withMVar` is defined slightly differently in `base`:
```
withMVar :: MVar a -> (a -> IO b) -> IO b
withMVar m io =
  mask $ \restore -> do
    a <- takeMVar m
    b <- restore (io a) `onException` putMVar m a
    putMVar m a
    return b
```

I believe both versions are correct, but there are many corner cases and I would rather copy that version and trust it instead of using slightly different manually defined one.

Reply:
Oh, btw, I confused `withMVar` and `modifyMVar`, but it doesn't really matter.